searx/user_settings.py

```diff
@@ -0,0 +1,239 @@
+from searx import settings
+from searx.languages import language_codes
+from searx.autocomplete import backends as autocomplete_backends
+from searx import utils
+
+
+class UnknownSetting(Exception):
+    pass
+
+
+class InvalidSetting(Exception):
+    pass
+
+
+class ValueErrorInvalidSetting(InvalidSetting):
+    pass
+
+
+class TypeErrorInvalidSetting(InvalidSetting):
+    pass
+
+
+class UserSettingsBase(object):
+    defaults = None
+    _collection_fields = {}
+
+    def __init__(self):
+        self._settings = dict()
+        self._is_empty = True
```

Comment:
`as_cookie_dict` now basically handles the serialization process. I don't think it would be a good idea to store values serialized. So I'm afraid we do need at least some algorithm here :(